src/plugins/core/typing_emoji_panel.ts

```diff
@@ -0,0 +1,188 @@
+    /**
+     * emojiStartPos
+     *
+     * position which corresponds to first char of
+     * the emoji name
+     *
+     * @example
+     * user wrote "hello :justatest: world"
+     * emojiStartPos would be 6
+     */
+    emojiStartPos: number;
+
+    /**
+     * typingEmojiPanel
+     *
+     * is the div which holds the list of emojis
+     */
+    typingEmojiPanel: HTMLDivElement;
+
+    currentEmoji: string;
+    caretPosition: number;
+    messageInp: HTMLInputElement;
+    isVisible: boolean;
+
+    constructor() {
+        super('typing_emoji_panel')
+
+        this.currentEmoji = '';
+        this.emojis = [];
+        this.layoutDiv = document.querySelector('.layout');
+        this.layoutDiv.insertAdjacentHTML('beforeend', `<div id="typing-emoji-panel"></div>`);
+        this.typingEmojiPanel = document.querySelector('#typing-emoji-panel');
+        this.messageInp = document.querySelector('#message-input');
+        this.caretPosition = 0;
+    }
+
+    hide(): void {
+        this.typingEmojiPanel.classList.remove('active');
+        this.isVisible = false;
+    }
+
+    show(): void {
+        this.typingEmojiPanel.classList.add('active');
+        this.isVisible = true;
+    }
+
+    onInit(): void {
+        this.messageInp.addEventListener('input', () => {
+            const emojis = this.typingEmojiPanel.querySelectorAll<HTMLDivElement>('[data-emoji-name]');
+
+            if (emojis.length > 0) {
+                this.currentEmoji = emojis[0].dataset.emojiName;
+            }
+        })
+        this.messageInp.addEventListener('keydown', (event) => this.onDocumentKey(event));
+        this.messageInp.addEventListener('keyup', () => this.update());
+        this.messageInp.addEventListener('click', () => this.update());
+        this.typingEmojiPanel.addEventListener('click', (e) => {
+            let target: HTMLElement = e.target as HTMLDivElement;
+
+            if (!target.dataset.emojiName) {
+                target = target.parentElement;
+            }
+
+            this.insertEmoji(target.dataset.emojiName);
+        });
+        document.addEventListener('click', (event) => this.onGlobalClick(event))
+        this.renderList();
+    }
+
+    onGlobalClick(event: MouseEvent) {
+        const clickTarget: HTMLElement = event.target as HTMLElement
+
+        if (!this.typingEmojiPanel.contains(clickTarget) && event.target != this.messageInp) {
+            this.hide()
+        }
+    }
+
+    onDocumentKey(e: KeyboardEvent): void {
+        if (e.key == 'ArrowUp') {
+            e.preventDefault();
+            this.selectPrevEmoji();
+        } else if (e.key == 'ArrowDown') {
+            e.preventDefault();
+            this.selectNextEmoji();
+        } else if (e.key == 'Tab' && this.isVisible) {
+            e.preventDefault();
+            this.insertEmoji(this.currentEmoji);
+        }
+
+        this.update();
+    }
+
+    selectNextEmoji() {
+        let index = this.emojis.indexOf(this.currentEmoji);
```

Comment:
Could be ``const index = ..`` the in a few other places too. At some point we should setup linting to force the usage of const if the variable is not mutated. Lets try to already use const so the diff later is not too big.